Top-level CMake

arnout opened the pull request:
For proper integration with build systems (OpenWrt and RDK-B) it is much
easier if we can build all of prplMesh in a single go with cmake,
without the need for maptools.py to build all components.

This PR moves the framework, common, controller and agent components to the top level.

Fixes #468.

arnout committed:
cmake: add top-level CMakeLists.txt for framework

For proper integration with build systems (OpenWrt and RDK-B) it is much
easier if we can build all of prplMesh in a single go with cmake,
without need for maptools.py to build all components.

Add a top-level CMakeLists.txt. For the time being, it only builds the
framework. The helper files move from framework/cmake to the top level.

Update maptools build.py to build the top-level instead of the framework
component.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
common: don't import bcl

The bcl library is imported with find_package as if it is an external
library. This complicates things without any discernable benefit.
Therefore, don't call find_package for it, and use 'bcl' instead of
'beerocks::bcl'.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
common: don't import btlvf

The btlvf library is imported with find_package as if it is an external
library. This complicates things without any discernable benefit.
Therefore, don't call find_package for it, and use 'btlvf' instead of
'beerocks::btlvf'.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
agent: build with top-level CMakeLists.txt

Add agent as a subdirectory in the top-level CMakeLists.txt and update
maptools accordingly.

Since agent is now built before framework and common are installed, we
can no longer use find_package for the internal libraries.

Remove the mapf:: and beerocks:: prefix for libraries that are now
internal instead of imported.

The cmake helpers in agent are no longer needed so remove them.

The top-level external-toolchain files can be used as is.

Update monitor to include bwl INTERFACE_INCLUDE_DIRECTORIES for bwl
headers (no need for target_link_libraries(bwl) since it is dynamically
loaded).

Update agent to include tp include bwl and bpl
INTERFACE_INCLUDE_DIRECTORIES for bwl and bpl headers for the same
reason.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>
Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout committed:
common: build with top-level CMakeLists.txt

Add common as a subdirectory in the top-level CMakeLists.txt and update
maptools accordingly.

Since common is now built before framework is installed, we can no
longer rely on find_program to find tlvf.py. Instead, define a custom
propert on the tlvf library to point to the source path of tlvf.py, and
use that property in btlvf.

Since common is now built before framework is installed, we can no
longer use the same find_package for MapfTransport. Change the
conditions to use MSGLIB to determine if we have a transport or not.

Remove the mapf:: prefix for libraries that are now internal instead of
imported.

Because the bcl package depends on tlvf and elpp, the order of
find_package calls becomes important. Therefore, move the tlvf and elpp
find_package calls before the one of bcl in agent and controller.

The cmake helpers in common are no longer needed so remove them.

The top-level external-toolchain files will be used. The only
significant difference was that framework set the default
TARGET_PLATFORM in its CMakeLists.txt, while common set it in
external-toolchain.cfg. Since relying on external-toolchain.cfg is a bad
idea, move the default from framework to top-level and remove it from
external-toolchain.cfg.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

tomereli reviewed (approved):
So far, so good. But this means that we can't build each component by its own anymore, doesn't it?

morantr commented:
I think we should wait before merge it until we will merge all the changes from the turris omnia.

arnout commented:
> I think we should wait before merge it until we will merge all the changes from the turris omnia.

OK for me - it indeed creates conflicts that have to be resolved. Then I can include agent and controller into the top-level build in the same PR as well.

And anyway, CI is currently failing on this so it can't be merged as is.

arnout committed:
common: bcl: cmake: remove redundant include

Since bwl links with bcl explicitly, the bcl include directories are
implied. Therefore, adding them to include_directories is redundant.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
common: bwl: use include <> instead of include ""

The bwl implementations use include statements like

  #include "../common/foo.h"

to include the bwl headers. Simplify this by adding the common directory
to the include path and use

  #include <foo.h>

Except for the CMakeLists.txt update, this is a mechanical change with

git grep -l '#include "../common/\([^"]*\)"' common/beerocks/bwl | \
   xargs sed -i 's%#include "../common/\([^"]*\)"%#include <\1>%'

This change will make it easier to move the common directory to another
place in a later commit.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
common: bwl: move public headers to include/

Similar like is done in other modules, put the public headers in the
include/ directory, with the same directory structure as where it is
installed.

This simplifies the installation command to just use DIRECTORY instead
of listing the files.

base_wlan_hal.cpp doesn't move, but it's #include is updated to use <>
instead of "" since the included file is no longer in the same
directory. Keep the <bwl/..> prefix to diffrenciate between include
files that can potentially have the same name.

While we're at it, also make the include directory a public include
directory, for the sake of internal consumers (of which there are none
at the moment).

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>
Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: btl: add build interface include dir

Add build interface include directories in CMakeLists.txt so that
target_link_libraries will include the btl.h header when used by other
modules.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
controller: build with top-level CMakeLists.txt

Add controller as a subdirectory in the top-level CMakeLists.txt and update
maptools accordingly.

Since controller is now built before framework and common are installed, we
can no longer use find_package for the internal libraries.

Remove the mapf:: and beerocks:: prefix for libraries that are now
internal instead of imported.

The cmake helpers in agent are no longer needed so remove them.

The top-level external-toolchain files can be used as is.

Move Findreadline.cmake to top level cmake directory.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
build: options for not building controller/agent

Add cmake options for skipping build for controller and/or agent.
This can be useful in several build configurations:

Framework only - for users interested only in the framework (AKA 1905
transport and local bus)

Framework + Agent - for saving storage space for extenders/repeaters
which do not need a controller

Framework + Controller - for cloud or proxy controllers with no actual
wifi hardware

Framework + Controller + Agent (default) - Full solution

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tools: maptools build.py align to single CMAKE

- Remove map_modules which is no longer needed since we use CMAKE options
for skipping build for controller and/or agent.

- Remove distclean option - simple clean is now sufficient

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

vitalybu commented on CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+endif()
+
+option (BUILD_AGENT "Build EasyMesh agent" ON)
+option (BUILD_CONTROLLER "Build EasyMesh agent" ON)
```
Incorrect option description. Should be "Build EasyMesh controller".

vitalybu commented on CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+endif()
+if (BUILD_CONTROLLER)
+add_subdirectory(controller)
+endif()
```
Missing new line at the end of the file

vitalybu commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -121,9 +121,9 @@ bool slave_thread::init()
-                                                 config.model, config.bridge_iface, &master_socket);
+        m_agent_ucc_listener = std::make_unique<agent_ucc_listener>(
+            config.ucc_listener_port, config.vendor, config.model, config.bridge_iface,
+            (Socket **)&master_socket);
```
A better solution is to change the type of the controller socket to `SocketClient` in the `agent_ucc_listener` class.
Already did it in the UGW Support branch, so can be fixed later.

tomereli replied:
ok, so once you merge the UGW Support branch you can override it.

vitalybu commented on common/beerocks/btl/CMakeLists.txt:
```diff
@@ -27,17 +23,19 @@ set(MODULE_PATH ${CMAKE_CURRENT_LIST_DIR})
         ${MODULE_PATH} 
         ${MODULE_PATH}/../../../external
-        $<TARGET_PROPERTY:beerocks::bcl,INTERFACE_INCLUDE_DIRECTORIES>
+        $<TARGET_PROPERTY:bcl,INTERFACE_INCLUDE_DIRECTORIES>
```
Why is this required anyway?
It makes sense for BWL and BPL as they are loaded dynamically, but for the BCL that it linked during build time, `target_link_libraries` should be enough, no?
If not, perhaps we should update the target to correctly set the INTERFACE_INCLUDE_DIRECTORIES.

tomereli replied:
You're right - it's not needed, and neither does the external (which no longer exists nor the MODULE_PATH

vitalybu commented on common/beerocks/bwl/brcm/ap_wlan_hal_brcm.h:
```diff
@@ -9,8 +9,8 @@
 
-#include "../common/ap_wlan_hal.h"
 #include "base_wlan_hal_brcm.h"
+#include <ap_wlan_hal.h>
```
I think that relative paths for files in the same module is better practice.
Can prevent issues when files from different modules have the same name.

This comment is relevant for all BPL/BWL files.

tomereli replied:
The idea is to be able to easily move the BWL/BPL to the framework so we don't maintain 2 different common libraries. I also agree with you that using only the file name is problematic - so how about a compromise, adding a `bwl` directory prefix? for example, `#include <bwl/base_wlan_hal.h>`

tomereli replied:
If you agree with this approach, I will add this change in a separate commit right after this one.

vitalybu replied:
You can keep the relative paths and still move the modules between the projects.
Relative paths are used only for including files from the same logical group of files (e.g. inside the BWL module)

tomereli replied:
OK, so I'll remove this commit. @arnout if you still think this is required, we change it later.

tomereli replied:
On second thought we decided to keep Arnout's change but to use relative paths to `<bwl/...>` `<bpl/,,,>` `<btl/,,` etc. So I will change this commit by Arnout accordingly.

vitalybu commented on controller/src/beerocks/master/CMakeLists.txt:
```diff
@@ -24,19 +20,12 @@ endif()
-    ${MODULE_PATH}/../../../external
-    $<TARGET_PROPERTY:beerocks::bcl,INTERFACE_INCLUDE_DIRECTORIES>
-    $<TARGET_PROPERTY:beerocks::btlvf,INTERFACE_INCLUDE_DIRECTORIES>
+    # $<TARGET_PROPERTY:beerocks::bcl,INTERFACE_INCLUDE_DIRECTORIES>
```
I guess it's not necessary after all...
Should be removed from all other CMake files as well.

Commit:
common: bcl: <bcl/..> instead of <beerocks/bcl/..>

Similar to what was done with bwl, now refactor bcl include directories
so that all includes use <bcl/..> prefix instead of the longer and
obsolete <beerocks/bcl/..>.

While we're at it - skip installing external header files for the BCL
since they are not used by anyone anymore. Removing the export
installation will come in a separate commit since it requires to remove
all dependencies exports as well.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: btl: remove unused export and files

Now that prplMesh is built in a single CMAKE project, there is no need
to install anything other than the library itself. No need for include
files, nor export, since no external component is expected to link with
btl.

While at it, also align with bwl and bcl convention to use <btl/btl.h>

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: bwl: remove unused export and files

Now that prplMesh is built in a single CMAKE project, there is no need
to install anything other than the library itself. No need for include
files, nor export, since no external component is expected to link with
btl.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
framework: bpl: remove unused export and files

Now that prplMesh is built in a single CMAKE project, there is no need
to install anything other than the library itself. No need for include
files, nor export, since no external component is expected to link with
bpl.

While at at, also align bpl to other libraries include directories so
that <bpl/..> is used instead of <..>.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
framework: remove unused export and files

Remove unused exports and installed files for the framework's transport
and common, since now that prplMesh is built in a single CMAKE project,
there is no need to install anything other than the libraries itself.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout reviewed (commented):
I don't agree with making btlvf a static lib.

1. The point of #469 is *not* to make everything static. Rather, the point is to make it *possible* to build static.
2. Linking a static library into a dynamic library only works if the static library is built with `-fPIC`. We happen to do that, but it's quite hidden.
3. This is really bad for code size: the btlvf (which is very large) will be linked into both libbml and the controller executable as weak symbols. LTO is not going to save is there.
4. It's not really needed. Even if btlvf is built as a shared library, that doesn't mean it has to be exported. Since bml is linked dynamically with btlvf, the dynamic linker will resolve it. So it is only needed if bml is built as a static library - but then this doesn't solve it, because the bml static library will *not* include btlvf so we still need to link with btlvf explicitly.
5. btlvf is no different from bcl, elpp, tvlf, so if we do this, the same should be done for those libraries.

So, I removed the last two commits and pushed again. That's what I will merge.

I pushed the last two commits as branch feature/clean-up-bml-export.

I'll make a new branch that does further cmake cleanups which will include most of that, except for the shared->static change.
